src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -362,45 +415,37 @@ protected final boolean hasMoreReferrals() throws NamingException {
      * with those of the current enumeration.
      */
     protected void update(AbstractLdapNamingEnumeration<? extends NameClassPair> ne) {
-        // Cleanup previous context first
-        homeCtx.decEnumCount();
-
-        // New enum will have already incremented enum count and recorded clnt
-        homeCtx = ne.homeCtx;
-        enumClnt = ne.enumClnt;
-
-        // Do this to prevent referral enumeration (ne) from decrementing
-        // enum count because we'll be doing that here from this
-        // enumeration.
-        ne.homeCtx = null;
-
-        // Record rest of information from new enum
-        posn = ne.posn;
-        limit = ne.limit;
-        res = ne.res;
-        entries = ne.entries;
-        refEx = ne.refEx;
-        listArg = ne.listArg;
-    }
-
-    @SuppressWarnings("removal")
-    protected final void finalize() {
-        cleanup();
+        try {
+            // Cleanup previous context first
+            homeCtx().decEnumCount();
+
+            // New enum will have already incremented enum count and recorded clnt
+            enumCtx.setHomeCtx(ne.enumCtx.homeCtx);
+            enumCtx.setEnumClnt(ne.enumCtx.enumClnt);
+
+            // 'this' and 'ne' now both refer to ne's homeCtx. 'this' will
+            // decrement homeCtx's enum count later (via cleanup() or Cleaner).
+            // Clear ne's reference to homeCtx so ne's Cleaner doesn't
+            // *also* decrement the count.
+            ne.enumCtx.setHomeCtx(null);
+
+            // Record rest of information from new enum
+            posn = ne.posn;
+            limit = ne.limit;
+            enumCtx.setRes(ne.enumCtx.res);
+            entries = ne.entries;
+            refEx = ne.refEx;
+            listArg = ne.listArg;
+        } finally {
+            // Ensure Cleaner does not run until after this method completes
```

Comment:
Ditto try/finally is unnecessary.

Reply:
> Ditto try/finally is unnecessary.

Perhaps, because there may be only one exit path. But style-wise it is still better to play safe for any possible future changes to this part of code. Memory fence is not needed though as explained above.